### PR TITLE
feature(planner): Decorrelate `EXISTS` subquery with non-equi condition

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Changelog

- New Feature

### CODE REVIEW

File: query/src/sql/exec/pipeline_builder.rs [hunk 1]
```diff
@@ -487,12 +487,35 @@ impl PipelineBuilder {
         mut child_pipeline: NewPipeline,
         pipeline: &mut NewPipeline,
     ) -> Result<()> {
+        let predicate = other_conditions.iter().cloned().fold(
```
Comment (andylokandy) [inline query/src/sql/exec/pipeline_builder.rs:490]: You can use `https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_reduce` instead.

**Review by andylokandy — APPROVED** [review]

File: query/src/sql/exec/pipeline_builder.rs (hunk 1 above)
Comment (leiysky) [reply query/src/sql/exec/pipeline_builder.rs:490]: > You can use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_reduce instead.

It's unstable feature, I don't want to enable it...

File: query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs [hunk 2]
```diff
@@ -469,11 +468,32 @@ impl ChainingHashTable {
                         {
                             let keys_ref = KeysRef::create(key.as_ptr() as usize, key.len());
                             let probe_result_ptr = table.hash_table.find_key(&keys_ref);
-                            if probe_result_ptr.is_none() {
-                                // No matched row for current probe row
-                                continue;
+                            match (probe_result_ptr, &self.other_predicate) {
```
Comment (sundy-li) [inline query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:471]: Should we also modify `result_blocks` ?

Comment (leiysky) [reply query/src/pipelines/new/processors/transforms/hash_join/chaining_hash_table.rs:471]: > Should we also modify `result_blocks` ?

We don't actually need to, this is the Semi/Anti join-specific behavior.

**Pull request merged**